### ISSUE TITLE
Training sample ratio on random forest for gamma/hadron separation

### ISSUE THREAD

**Issue by SeiyaNozaki** [cta-observatory/cta-lstchain#701]
Now `test_size` for random forest is 0.2.
https://github.com/cta-observatory/cta-lstchain/blob/master/lstchain/reco/dl1_to_dl2.py#L258

If I understand correctly, this `test_size` is not for the actual test, but a training sample for gamma/hadron separation.
So when it is 0.2, 80% of gamma MC is used for energy/reco position temporal RF training before gamma/hadron separation RF, and 20% gamma MC+100% proton MC is used for gamma/hadron separation.

In my feeling, 0.2 is a bit small... so it would be better to use 0.5 or so, or add an option to use user-selectable values in `lstchain_dl1_to_dl2.py`.
What do you think?

**Comment by maxnoe** [comment]
There is also no real reason to not use the same gammas for gamma/hadron and energy regressor training, if one is not used as input as the other.

**Comment by rlopezcoto** [comment]
> There is also no real reason to not use the same gammas for gamma/hadron and energy regressor training, if one is not used as input as the other.

I think it is slightly different how it works: 
1) you use all the gammas for energy/direction regressor training ---> this is what we store in the models that are later applied to the data
2) you use 80% of the gammas (`train`) to train an energy/disp regressor that is applied to the `testg` (20% of the gammas) to estimate their energy/direction because this info is used in the gamma/hadron separation train (see [classification features](https://github.com/cta-observatory/cta-lstchain/blob/20fbf9b5cf0a98e32a8838915e383697f3822a50/lstchain/data/soft_cut_config.json#L100-L102))

I agree with @SeiyaNozaki that we could increase this, but I recall @misabelber doing some tests long ago with previous productions and she may comment on this.

**Comment by moralejo** [comment]
> I think it is slightly different how it works:
> 
> 1. you use all the gammas for energy/direction regressor training ---> this is what we store in the models that are later applied to the data
> 2. you use 80% of the gammas (`train`) to train an energy/disp regressor that is applied to the `testg` (20% of the gammas) to estimate their energy/direction because this info is used in the gamma/hadron separation train (see [classification features](https://github.com/cta-observatory/cta-lstchain/blob/20fbf9b5cf0a98e32a8838915e383697f3822a50/lstchain/data/soft_cut_config.json#L100-L102))
>

Are you sure this is done like that @rlopezcoto ?  We should never apply a different RF to data and test MC. What is described above would not be terribly harmful because you never apply a model to an MC event which has been part of its training sample, but it will nevertheless introduce one more difference between data and MC.

**Comment by misabelber** [comment]
Hi all:
- @moralejo As far as I know and regarding the tests that I did long time ago, (and I think this procedure was largely discussed back then, shown in the notebooks and so on) it is how @rlopezcoto describes. The energy/disp regressors were trained with ALL the diffuse gammas. These models are then stored and used to reconstruct energy and disp for data, or for testing MC in case is needed. To be honest, now I'm not sure if this is the case being carried out in LP right now, because I can see that all the dl1 files are being merged into training and testing sets, so probably only the "training" set of diffuse gammas is used for the energy/disp regression. Under my understanding, we could be using all the diffuse gammas for training, since for testing and producing performance results, we only use pointlike gammas, so it makes no sense to divide each of them into two sets.   
- Because for gamma/hadron separation we need the reconstructed energy and disp as parameters, we take the same diffuse gammas we used to train the regressors, then divide them in training and test sets and then train other "smaller" regressors that we apply to these test diffuse gammas and to the protons, which afterwards are used for training the gh classifier. 
- These smaller regressors are only used to calculate energy and disp of the training sets of the classifier, but for giving any results and performance, with MC or data, we apply the firstly trained regressors, trained with a larger set of diffuse gammas.  

I think that, as @SeiyaNozaki suggested, the proportion for training these smaller regressors should be 50%, since the "test" gamma set is used for training the classifier, so its better not to lose that many events (that was the proportion originally used in the paleolithic versions of lstchain).

**Comment by maxnoe** [comment]
> Under my understanding, we could be using all the diffuse gammas for training, since for testing and producing performance results, we only use pointlike gammas, so it makes no sense to divide each of them into two sets.

To produce Full-Enclosure IRFs we also need a diffuse gamma test dataset


> Because for gamma/hadron separation we need the reconstructed energy and disp as parameters, we take the same diffuse gammas we used to train the regressors, then divide them in training and test sets and then train other "smaller" regressors that we apply to these test diffuse gammas and to the protons, which afterwards are used for training the gh classifier.

This is not ok! You then use a different estimator in the training than for application. This can lead to mismatches (not between data and mc, but between the model used to predict energy in the training and the one used in application).

You must only train one energy model on an independent gamma diffuse sample and this is the model that will be used for everything (predicting on data and predicting the training data)

This all can be avoided by not using one learner's output as input for the other. Have you actually tested that the g/h classifier gets significantly better when provided with the estimated energy as opposed to just allowing a bit more complex model using only the "normal" features?

> These smaller regressors are only used to calculate energy and disp of the training sets of the classifier, but for giving any results and performance, with MC or data, we apply the firstly trained regressors, trained with a larger set of diffuse gammas.

Again, this uses different models on during the training process and the application process for the g/h classifier. This must not be the case.

**Comment by rlopezcoto** [comment]
thanks @misabelber for the clarification.
I think @moralejo is right and this may be introducing some mismatch between data and MC (although not dramatically). One solution could be:
- train energy/disp RF with 50% of the training gammas
- apply these models to the other 50% of the training gammas, that will later be used to train the RF for separation

@misabelber in la Palma, there is an additional "test" MC dataset that is needed, to which we apply all the models and is used to perform the IRF calculation, we are probably a bit confused due to the naming inside the lstchain code.

**Comment by vuillaut** [comment]
Hi

This is what is happening:
![image](https://user-images.githubusercontent.com/4263646/116054585-1990af00-a67c-11eb-9f23-c285d3f6d4a6.png)


See the code:
https://github.com/cta-observatory/cta-lstchain/blob/5032edb033a3d01bc6db45b930ad62d2df3a302a/lstchain/reco/dl1_to_dl2.py#L331

**Comment by vuillaut** [comment]
> Hi
> 
> This is what is happening:
> ![image](https://user-images.githubusercontent.com/4263646/116054585-1990af00-a67c-11eb-9f23-c285d3f6d4a6.png)
> 
> See the code:
> https://github.com/cta-observatory/cta-lstchain/blob/5032edb033a3d01bc6db45b930ad62d2df3a302a/lstchain/reco/dl1_to_dl2.py#L331

Note:
Point like IRFs are also produced from point-like test gammas.

**Comment by vuillaut** [comment]
I think the procedure is correct.
It is probably not optimised ratio-wise though.

**Comment by rlopezcoto** [comment]
thanks @vuillaut for the scheme. A couple of comments:
- I think calling the 20% used to _train_ the classifier "test" is the one that was a bit confusing.
- The point raised by @moralejo and @maxnoe is that the models that are applied to the data and the models that are applied to the MC that is used to train the separation are different(Energy/disp and tmp energy/tmp disp) in your blue boxes, and this may induce data/MC mismatches.

**Comment by vuillaut** [comment]
> thanks @vuillaut for the scheme. A couple of comments:
> 
> * I think calling the 20% used to _train_ the classifier "test" is the one that was a bit confusing.

Yes, I agree.

> * The point raised by @moralejo and @maxnoe is that the models that are applied to the data and the models that are applied to the MC that is used to train the separation are different(Energy/disp and tmp energy/tmp disp) in your blue boxes, and this may induce data/MC mismatches.

Yes but that is not true if you look at the schema, the same model is applied to the MC test dataset to produce the IRFs and to real data.

**Comment by moralejo** [comment]
Thanks @misabelber, @vuillaut et al for the explanations.

So it is not data and test MC that use different RFs.
It is only the g/h separation training sample that  gets an Ereco and direction_reco from different "tmp" RFs. Probably not a big deal, then. At most it would result in "not perfect" g/h training (and I doubt it is even noticeable). But it won't produce  any data vs MC bias, because both are processed identically.

We should understand, as @maxnoe says, whether we really need the reco E and direction as g/h sep RF inputs or not. Perhaps it is not worth the complication.

Then, on the original topic of this issue: @SeiyaNozaki I think the limitation of the g/h training is anyway on the proton side, because of the low fraction of g-like hadrons. But most gammas are gamma-like, so it is not clear to me that you can gain much by just increasing the gamma statistics. Do you have some indication that you have too few gammas for training?

**Comment by maxnoe** [comment]
But the g/h model itself has a mismatch in the predicted energy. The output of the smaller model might be different to the output of the model just in data.

It's not a problem that the IRFs will be wrong, just that the results might be suboptimal since there is a difference between the training data and application data for the g/h classifier.

**Comment by rlopezcoto** [comment]
> Yes but that is not true if you look at the schema, the same model is applied to the MC test dataset to produce the IRFs and to real data.

according to the scheme, the energy/direction estimation used for IRF calculation (and also applied to real data) are "Energy/disp" models, right? 
To the MC used to train the classifier, there are different models used "tmp energy/tmp disp", and this is what could induce  mismatches between models

**Comment by misabelber** [comment]
Hi all, thanks for the clarifications!
As @moralejo says, because both data and test MC are processed the same way, the effect should be maybe seen in the classification, but not as a mismatch between real or MC data.
I think this procedure was made this way, because the regressors highly benefit form having more events in the training. I guess we could study the effect on the classification implementing the solution suggested by @rlopezcoto and see if its really worth it to change the procedure ?

**Comment by maxnoe** [comment]
> I think this procedure was made this way, because the regressors highly benefit form having more events in the training.

The more they benefit from more events, the higher the mismatch we are talking about here.

**Comment by vuillaut** [comment]
The other possibility would be to:

1. split train gammas in sub_train/validation
2. train reg on sub_train
3. infer reg on validation
4. train classifier on validation

In this case we use the same model for final regression and to train the classifier, but we decrease the performance of the final regressor because we use less data for its training.
So we **potentially** trade loss in regression for gain in classification.

The important word is *potentially*, the conclusion would be highly dependent on the total number of events used to train and in the sub_train/validation repartition.

**Comment by vuillaut** [comment]
But if the `tmp_reg` is not doing too bad a job compared to the final one (which should be the case, since we use 80% of the train dataset), I suspect a very limited gain.

Using regression features as input for the classifier is what is done in H.E.S.S. and MAGIC, no?
Any good biblio to see if comparative studies have been done and how people converged to that procedure in the past?

**Comment by rlopezcoto** [comment]
> 1. split train gammas in sub_train/validation
> 2. train reg on sub_train
> 3. infer reg on validation
> 4. train classifier on validation

@vuillaut I think this is exactly what is proposed here:
> * train energy/disp RF with 50% of the training gammas
> * apply these models to the other 50% of the training gammas, that will later be used to train the RF for separation

right?

**Comment by SeiyaNozaki** [comment]
Thank you all for the feedback!

> Then, on the original topic of this issue: @SeiyaNozaki I think the limitation of the g/h training is anyway on the proton side, because of the low fraction of g-like hadrons. But most gammas are gamma-like, so it is not clear to me that you can gain much by just increasing the gamma statistics. Do you have some indication that you have too few gammas for training?

At least for source-dependent analysis, I found more statistics of gamma MC would give better classification performance, especially for low-energy (< 100 GeV). It would also make gammaness of a bit more proton MC events higher, but a very small fraction of all events. 
![image](https://user-images.githubusercontent.com/31307823/116063041-76618900-a6bf-11eb-8312-607d337bf2cc.png)

**Comment by vuillaut** [comment]
> > 1. split train gammas in sub_train/validation
> > 2. train reg on sub_train
> > 3. infer reg on validation
> > 4. train classifier on validation
> 
> @vuillaut I think this is exactly what is proposed here:
> 
> > * train energy/disp RF with 50% of the training gammas
> > * apply these models to the other 50% of the training gammas, that will later be used to train the RF for separation
> 
> right?

Ha yes, sorry I overlooked it.

But I insist, there is *no* mismatch between MC and real data, just a potential suboptimal classifier performance.

**Comment by vuillaut** [comment]
> Thank you all for the feedback!
> 
> > Then, on the original topic of this issue: @SeiyaNozaki I think the limitation of the g/h training is anyway on the proton side, because of the low fraction of g-like hadrons. But most gammas are gamma-like, so it is not clear to me that you can gain much by just increasing the gamma statistics. Do you have some indication that you have too few gammas for training?
> 
> At least for source-dependent analysis, I found more statistics of gamma MC would give better classification performance, especially for low-energy (< 100 GeV). It would also make gammaness of a bit more proton MC events higher, but a very small fraction of all events.
> ![image](https://user-images.githubusercontent.com/31307823/116063041-76618900-a6bf-11eb-8312-607d337bf2cc.png)

Hi Seiya, thanks for that figure.

Did you check the classification performance with an AUC score, or even final sensitivity?
From your figure, with a cut at gammaness=0.6, there might actually be no difference in performances, the number of gammas above the threshold might be similar and there would even be more protons.

**Comment by moralejo** [comment]
If you change the statistics of just one of the species, the distributions of gammaness will shift, for both species, but this does not necessarily imply a gain. Please @SeiyaNozaki do the test @vuillaut proposes.

**Comment by SeiyaNozaki** [comment]
Thank you for the comments @vuillaut @moralejo !
Actually I didn't do additional checks, so I will do it following your comments.

**Comment by satoshifukami0115** [comment]
Sorry for jumping in, but I performed this kind of test a bit in the past, so let me share the results (sorry it includes some Japanese).
You can see how gammaness distributions change with train sample ratio of g&h. No event cuts. (all the events after image cleaning). lstchain was v0.6.1 I guess.
AUC was not changed much. I could not see any trend, although I did try how sensitivity like eff(g)/sqrt(eff(p)) changes with the ratio.

<img width="946" alt="Screen Shot 2021-05-09 at 23 58 34" src="https://user-images.githubusercontent.com/39740892/117576908-86975080-b122-11eb-87f3-162937d41934.png">
